be/src/vec/common/sort/sorter.cpp

```diff
@@ -17,48 +17,154 @@
+    num_rows_ += rows;
+    return Status::OK();
+}
+
+Status MergeSorterState::_build_merge_tree_not_spilled(const SortDescription& sort_description) {
+    for (const auto& block : sorted_blocks_) {
+        cursors_.emplace_back(block, sort_description);
+    }
+
+    if (sorted_blocks_.size() > 1) {
+        for (auto& cursor : cursors_) priority_queue_.push(MergeSortCursor(&cursor));
+    }
+    return Status::OK();
+}
+
+Status MergeSorterState::build_merge_tree(const SortDescription& sort_description) {
+    _build_merge_tree_not_spilled(sort_description);
+
+    if (spilled_sorted_block_streams_.size() > 0) {
+        if (sorted_blocks_.size() > 0) {
+            BlockSpillWriterUPtr spill_block_wirter;
+            RETURN_IF_ERROR(ExecEnv::GetInstance()->block_spill_mgr()->get_writer(
+                    spill_block_batch_size_, spill_block_wirter));
+
+            LOG(WARNING) << "spill to disk: build merge tree, in memory sorted blocks: "
+                         << sorted_blocks_.size();
+            if (sorted_blocks_.size() == 1) {
+                RETURN_IF_ERROR(spill_block_wirter->write(sorted_blocks_[0]));
+            } else {
+                bool eos = false;
+
+                // merge blocks in memory and write merge result to disk
+                while (!eos) {
+                    merge_sorted_block_.clear_column_data();
+                    _merge_sort_read_not_spilled(spill_block_batch_size_, &merge_sorted_block_,
+                                                 &eos);
+                    RETURN_IF_ERROR(spill_block_wirter->write(merge_sorted_block_));
+                }
+            }
+            spilled_sorted_block_streams_.emplace_back(spill_block_wirter->get_id());
+            spill_block_wirter->close();
+        }
+        RETURN_IF_ERROR(_merge_spilled_blocks(sort_description));
+    }
+    return Status::OK();
 }
 
 Status MergeSorterState::merge_sort_read(doris::RuntimeState* state,
                                          doris::vectorized::Block* block, bool* eos) {
-    size_t num_columns = sorted_blocks[0].columns();
+    Status status = Status::OK();
+    if (is_spilled_) {
+        status = merger_->get_next(block, eos);
+        LOG(WARNING) << "spill to disk: merger_->get_next, eos: " << *eos;
+    } else {
+        LOG(WARNING) << "spill to disk: _merge_sort_read_not_spilled, eos: " << *eos;
+        if (sorted_blocks_.empty()) {
+            *eos = true;
+        } else if (sorted_blocks_.size() == 1) {
+            if (offset_ != 0) {
+                sorted_blocks_[0].skip_num_rows(offset_);
+            }
+            block->swap(sorted_blocks_[0]);
+            *eos = true;
+        } else {
+            status = _merge_sort_read_not_spilled(state->batch_size(), block, eos);
+        }
+    }
+    return status;
+}
+
+Status MergeSorterState::_merge_sort_read_not_spilled(int batch_size,
+                                                      doris::vectorized::Block* block, bool* eos) {
```

Comment:
Does this method could return not OK status? And I find you did not check the status.